ユーザーアイコンアップロード機能を実装

yama998877 opened the pull request:
ActiveStorageを使用。
`configure_permitted_parameters`の`keys`にiconを追加。
`with_attached_icon`を追加して、N+1問題を解消。
Userモデルに`icon`を定義。
`_profile_fields.html.erb`にアイコンのファイル入力ボックスを追加。
ユーザ一覧と詳細画面にリサイズされたアイコン画像が表示されるように変更。
リサイズに`mini_magick`を使うように設定。
アイコンのキーと翻訳文を追加。

yama998877 committed:
ユーザーアイコンアップロード機能を実装

ActiveStorageを使用。
`configure_permitted_parameters`の`keys`にiconを追加。
`with_attached_icon`を追加して、N+1問題を解消。
Userモデルに`icon`を定義。
`_profile_fields.html.erb`にファイル入力ボックスを追加。
ユーザ一覧と詳細画面にリサイズされたアイコン画像が表示されるように変更。
リサイズに`mini_magick`を使うように設定。
アイコンのキーと翻訳文を追加。

JunichiIto commented on config/application.rb:
```diff
@@ -18,5 +18,8 @@ class Application < Rails::Application
     # config.eager_load_paths << Rails.root.join("extras")
+
+    # 画像のリサイズで使用するGemの指定
+    Rails.application.config.active_storage.variant_processor = :mini_magick
```
vipsではなくmini_magickを使ったのには何か理由がありますか？

yama998877 replied:
> ImageMagickは、libvipsに比べて知名度が高く普及も進んでいます。

と[Railsガイド](https://railsguides.jp/v8.0/active_storage_overview.html#%E8%A6%81%E4%BB%B6)にあったので何かエラーがあったときに解決しやすいのはmini_magickかなと思って使用しました🙏

JunichiIto replied:
なるほどなるほど！たしかにそう書いてありますね。

ただ、この文言に続いてこんなふうに書いてあります。

> しかしlibvipsは10倍高速かつメモリ消費も1/10です。JPEGファイルの場合、libjpeg-devをlibjpeg-turbo-devに置き換えると2〜7倍高速になります。

これは「たしかにA。しかしB」の構文になっており、Railsガイドの意図としてはBを推したい、という意図があります。

実際、Rails 7からはvipsがデフォルトになっています。

https://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#active-storage-default-variant-processor-changed-to-vips

というわけで、「絶対こっち！」という強い理由がない限りはvipsを使った（つまりvariant_processorは指定せずにデフォルトを使う）方がベターだと思います〜。

yama998877 replied:
> これは「たしかにA。しかしB」の構文になっており、Railsガイドの意図としてはBを推したい、という意図があります。

構文になっていることに気づきませんでした🙇‍♂️
ImageMagickでないと難しい画像処理がないのでデフォルトのvipsを使うように変更しました🙏

JunichiIto reviewed (commented):
1点質問しました〜。

yama998877 committed:
画像のリサイズに使うGemの変更

mini_magickからruby-vipsに変更。
application.rbの`config.active_storage.variant_processor`を必要がなくなったので削除。

JunichiIto reviewed (approved):
確認しました！OKです〜🙆‍♂️
